src/sql/mod.rs

```diff
@@ -12,34 +12,43 @@ pub(crate) mod escape;
 mod ser;
 
 #[derive(Clone)]
-pub(crate) enum SqlBuilder {
-    InProgress(Vec<Part>),
+pub(crate) enum SqlBuilder<'a> {
+    InProgress(Vec<Part<'a>>),
     Failed(String),
 }
 
 #[derive(Clone)]
-pub(crate) enum Part {
+pub(crate) enum Part<'a> {
     Arg,
     Fields,
-    Text(String),
+    Str(&'a str),
+    String(String),
 }
 
-impl SqlBuilder {
-    pub(crate) fn new(template: &str) -> Self {
-        let mut iter = template.split('?');
-        let prefix = String::from(iter.next().unwrap());
-        let mut parts = vec![Part::Text(prefix)];
+impl<'parts> SqlBuilder<'parts> {
+    pub(crate) fn new(template: &'parts str) -> Self {
+        let mut parts = Vec::new();
+        let mut rest = template;
+        while let Some(idx) = rest.find('?') {
+            if rest[idx + 1..].starts_with('?') {
```

Comment:
What if we do `client.query("?")`? Or panic is [okay](https://github.com/ClickHouse/clickhouse-rs/issues/60#issuecomment-1546850356)?

Reply:
for `client.query("?")` `rest[idx + 1..]` here will be `""`, & `starts_with` will return `false`

Reply:
You are right, sorry.